File: baekjoon/1780_종이의_개수.py
import sys
sys.stdin = open('1780.txt', 'r')

import math

def go_into(s_row, s_col, log_size):
    size = 3 ** log_size
    start = mat[s_row][s_col]
    is_same = True
    for i in range(s_row, s_row + size):
        for j in range(s_col, s_col + size):
            if mat[i][j] != start:
                is_same = False
                break
        if not is_same:
            break
    if is_same:
        count[start] += 1
    else:
        next_size = 3 ** (log_size - 1)
        for i in range(3):
            for j in range(3):
                go_into(s_row + i * next_size, s_col + j * next_size, log_size - 1)


N = int(input())

mat = [0] * N

for i in range(N):
    aline = list(map(int, input().split()))
    mat[i] = aline

n = round(math.log(N, 3))
count = {-1: 0, 0: 0, 1: 0}

go_into(0, 0, n)

for value in count.values():
    print(value)

File: baekjoon/1780_종이의_개수.cpp
#include <iostream>
#include <math.h>
#include <typeinfo>

using namespace std;

int N;
int mat[2200][2200];
int counts[3] = {0};

double intlog(double base, double x)
{
    return (log(x) / log(base));
}

void go_into(int s_row, int s_col, int log_size)
{
    int size = (int) pow(3, log_size);
    int start = mat[s_row][s_col];
    bool is_same = true;

    for (int i = s_row; i < s_row + size; i++)
    {
        for (int j = s_col; j < s_col + size; j++)
        {
            if (mat[i][j] != start)
            {
                is_same = false;
                break;
            }
        }
        if (!is_same)
        {
            break;
        }
    }
    if (is_same)
        counts[start + 1]++;
    else
    {
        int next_size = (int) pow(3, log_size - 1);
        for (int i = 0; i < 3; i++)
        {
            for (int j = 0; j < 3; j++)
                go_into(s_row + i * next_size, s_col + j * next_size, log_size - 1);
        }
    }

}

int main()
{
    cin >> N;

    for (int i = 0; i < N; i ++)
    {
        for (int j = 0; j < N; j ++)
            cin >> mat[i][j];
    }

    int n = (int) floor(intlog(3, N) + 0.5);
    // cout << typeid(n).name() << endl;
    cout << "n: " << n << endl;

    go_into(0, 0, n);

    for (int i = 0; i < 3; i ++)
        cout << counts[i] << endl;

    return 0;
}